framework/utils/randomUtils.py

```diff
@@ -84,107 +86,156 @@ def testSampling(self, n=1e5):
+  if engine is None:
+    if stochasticEnv == 'crow':
+      distStochEnv.seedRandom(value)
+      engine=crowStochEnv
+    elif stochasticEnv == 'numpy':
+      replaceGlobalEnv=True
+      global npStochEnv
+      engine = npStochEnv
+
+  if isinstance(engine, np.random.RandomState):
+    engine = np.random.RandomState(value)
+  elif isinstance(engine, findCrowModule('randomENG').RandomClass):
+    engine.seed(value)
     if seedBoth:
       np.random.seed(value+1) # +1 just to prevent identical seed sets
+      # not sure if this loop shold one level up
+  if stochasticEnv== 'numpy' and replaceGlobalEnv:
+    #global npStochEnv
+    npStochEnv= engine
   print('randomUtils: Global random number seed has been changed to',value)
 
-def random(dim=1,samples=1,keepMatrix=False):
+def random(dim=1,samples=1,keepMatrix=False,engine=None):
   """
     Function to get a single random value, an array of random values, or a matrix of random values, on [0,1]
     @ In, dim, int, optional, dimensionality of samples
     @ In, samples, int, optional, number of arrays to deliver
     @ In, keepMatrix, bool, optional, if True then will always return np.array(np.array(float))
+    @ In, engine, instance, optional, random number generator
     @ Out, vals, float, random normal number (or np.array with size [n] if n>1, or np.array with size [n,samples] if sampels>1)
   """
+  ## choose an engine if it is none
+  if engine is None:
+    if stochasticEnv == 'numpy':
+      engine = npStochEnv
+    elif stochasticEnv == 'crow':
+      engine = crowStochEnv
   dim = int(dim)
   samples = int(samples)
-  if stochasticEnv == 'numpy':
-    vals = npStochEnv.rand(samples,dim)
-  else: #crow
+  if isinstance(engine, np.random.RandomState):
+    vals = engine.rand(samples,dim)
+  elif isinstance(engine, findCrowModule('randomENG').RandomClass):
     vals = np.zeros([samples,dim])
     for i in range(len(vals)):
       for j in range(len(vals[0])):
-        vals[i][j] = crowStochEnv.random()
+        vals[i][j] = engine.random()
+  else:
+    raise TypeError('Engine type not recognized! {}'.format(type(eng)))
   # regardless of stoch env
   if keepMatrix:
     return vals
   else:
     return _reduceRedundantListing(vals,dim,samples)
 
-def randomNormal(dim=1,samples=1,keepMatrix=False):
+def randomNormal(dim=1,samples=1,keepMatrix=False,engine=None):
   """
     Function to get a single random value, an array of random values, or a matrix of random values, normally distributed
     @ In, dim, int, optional, dimensionality of samples
     @ In, samples, int, optional, number of arrays to deliver
     @ In, keepMatrix, bool, optional, if True then will always return np.array(np.array(float))
+    @ In, engine, instance, optional, random number generator
     @ Out, vals, float, random normal number (or np.array with size [n] if n>1, or np.array with size [n,samples] if sampels>1)
   """
+  ## choose an engine if it is none
+  if engine is None:
```

Comment:
It looks like these 4 lines get repeated in almost every function; can we write a method to extract these four lines, and maybe the `TypeError` check, to a separate method?

Reply:
Added the get engine method for the engine is none. Did not change the TypeError check since it almost does the same thing?